Update to latest jest

ismay opened the pull request:
Wanted to see what updating jest to the latest version would entail for d2. Brought with it a couple of changes:

- Updated to latest babel to match jest 24's version of babel.
- The Model module was overwriting the prototype of Model with instantiated ModelBase class, which was failing with the new babel version and isn't recommended. Changed it to an extend, which broke tests. Refactored conservatively to pass tests again. Will need to test thoroughly with existing app(s) to make sure there are no regressions. I'm not sure why it was done this way, so we'll have to see if we can find out why.
- Used snapshots for error message checking to simplify changing error messages (as they had changed).

Todo

- [x] Refactored `done()` callback to use `expect.assertions()`
- [x] Check warnings being logged in the console (only ones remaining are fine)
- [x] The systemconfiguration snapshots contain undefined in two cases

To be discussed in review

- [ ] Existing tests should be reviewed, there are a lot of tests that don't seem that useful. There's a lot of mocking that makes it difficult to see what's actually being tested. Chances are that the library is actually broken, but that the tests are passing because of the mocks (which don't seem to be using a lot of actual module code, with [jestgenmockfrommodulemodulename](https://jestjs.io/docs/en/jest-object#jestgenmockfrommodulemodulename) for example). Plus, the tests seem to be testing the private structure of the library a lot of the time, and not sticking to targeted unit tests or the public API. Which means that internal structural changes require identical changes to the tests, which is cumbersome and not necessary. But that's probably better dealt with in a new PR.
- [ ] There's one test that's currently ignored (https://github.com/dhis2/d2/pull/190/files#diff-4c9d130892a8a64bf3d48aaf6ecbd5f1R74), this is because extending ModelBase means this property has become writeable (because js looks for the writeable flag on the parent, even though it'll write on the child). We'll have to check whether that's a problem. Refactoring this area the way I've currently done it seemed like the approach that stayed closed to the original design.

varl commented on src/model/ModelBase.js:
```diff
@@ -303,4 +302,4 @@ class ModelBase {
 }
 
-export default new ModelBase();
+export default ModelBase;
```
This seems more correct, and I think this might be why `Model` set the `ModelBase` object exported here on its own prototype, so it should be safe ... I think.

varl commented:
I've added the string following string to the files which are **not** test files and might have functional consequences for the next reviewer:

>As context for the next reviewer: this is a non-test change.

It's not meant to hilight any problems, just as a bookmark for non-test changes.

ismay commented:
As to what we were talking about @varl, apparently to specify a parent class, you do set an instance on `.prototype`, and then assign the constructor function to `prototype.constructor`:

> The Manager and WorkerBee definitions show the difference in how to specify the next object higher in the inheritance chain. In JavaScript, you add a prototypical instance as the value of the prototype property of the constructor function, then override the prototype.constructor to the constructor function. You can do so at any time after you define the constructor. In Java, you specify the superclass within the class definition. You cannot change the superclass outside the class definition.
> - https://developer.mozilla.org/bm/docs/Web/JavaScript/Guide/Details_of_the_Object_Model

So what was done here seems functionally the same as an extend to me.

ismay commented:
When using this branch of the package with the maintenance app, fetching data works for me:

<img width="1278" alt="screenshot 2019-02-25 at 10 22 18" src="https://user-images.githubusercontent.com/7355199/53327096-6a16f100-38e7-11e9-8097-9c6e232160cd.png">

<img width="1280" alt="screenshot 2019-02-25 at 10 22 37" src="https://user-images.githubusercontent.com/7355199/53327104-70a56880-38e7-11e9-997f-a855ace214eb.png">

varl commented:
>So it wasn't as weird as I thought. What was done here seems functionally the same as an extend.

Though it's still weird going for the ES5 syntax while implementing ES6 classes everywhere. Threw me for a loop.

>In JavaScript, you add a prototypical instance as the value of the prototype property of the constructor function, then override the prototype.constructor to the constructor function.

Interesting, so I guess `extends` has to do this behind the scene anyway. At least we don't have to worry about that it was done as a memory efficiency optimization.

Mohammer5 commented on src/model/ModelValidation.js:
```diff
@@ -3,6 +3,15 @@ import Logger from '../logger/Logger';
 
+function extractValidationViolations(webmessage) {
+    if (webmessage.response && webmessage.response.errorReports) {
+        return webmessage.response.errorReports;
```
Is this a rejected promise already? If not, then this should be wrapped inside a newly created, rejected promise as well, otherwise the return value types will differ.

ismay replied:
Yeah it is, that's actually why I'm returning a Promise in the other case as well. Let me see if I can find where this promise is created..

ismay replied:
Actually I was wrong, it's not a Promise, it's an object:

```javascript
[ { message: 'Required property missing.', property: 'name' } ]
```

Wrapping it in a `Promise.resolve` doesn't break tests, but I'll check it some more. See what's going on there.

ismay replied:
So, after looking at it, this whole method seems pretty unintuitive. To address your original question here, we don't necessarily have to return the same types as `then` will wrap the return value right? Correct me if I'm wrong. However, as to the weird things I found:

- [This line for example](https://github.com/dhis2/d2/blob/4d9475869a15c82738f4d4c1105ea8e625b63752/src/model/ModelValidation.js#L53) seems pretty pointless to me.
- And [here](https://github.com/dhis2/d2/blob/4d9475869a15c82738f4d4c1105ea8e625b63752/src/model/ModelValidation.js#L58) webmessage is passed to `.reject`, where I'd usually expect an error.
- Also [here](https://github.com/dhis2/d2/blob/4d9475869a15c82738f4d4c1105ea8e625b63752/src/model/ModelValidation.js#L6) the error is caught and resolves succesfully when there are error reports, seems counterintuitive, but I'm not sure what's going on here exactly. 

Seems like the control flow is a bit messed up here. But I think this is something we should address in a separate PR. I'll make some changes and ping you when I've pushed them.

ismay commented:
@Mohammer5 I've changed two minor things. [here](https://github.com/dhis2/d2/pull/190/files#diff-91881907882acbeb6acb67c4848b4dc9R6) I'm throwing an error since `then` will reject with the error automatically.

I've removed the unnecessary catch [here](https://github.com/dhis2/d2/pull/190/files#diff-91881907882acbeb6acb67c4848b4dc9L51), wasn't doing anything. Other than that it seems like things for a separate issue. Let me know what you think and I'll merge it if you agree.

Mohammer5 commented:
Approved already :)
